crates/red_knot_python_semantic/resources/mdtest/metaclass.md

```diff
@@ -0,0 +1,130 @@
+## Default
+
+```py
+class C: ...
+
+reveal_type(C.__class__)  # revealed: Literal[type]
+```
+
+## Basic
+
+```py
+class A: ...
+class B(metaclass=A): ...
+
+reveal_type(B.__class__)  # revealed: Literal[A]
+```
+
+## Linear inheritance (1)
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C(B): ...
+
+reveal_type(C.__class__)  # revealed: Literal[A]
+```
+
+## Linear inheritance (2)
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(B): ...
+
+reveal_type(C.__class__)  # revealed: Literal[A]
+```
+
+## Conflict (1)
+
+The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its
+bases.
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C: ...
+class D(metaclass=C): ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `Literal[C]` and `Literal[A]` are not compatible"
+class E(D, B): ...
+
+reveal_type(E.__class__)  # revealed: Unknown
+```
+
+## Conflict (2)
+
+The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its
+bases.
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C: ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `Literal[C]` and `Literal[A]` are not compatible"
+class E(B, metaclass=C): ...
+
+reveal_type(E.__class__)  # revealed: Unknown
+```
+
+## Inheritance (1)
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C(metaclass=A): ...
+class D(B, C): ...
+
+reveal_type(D.__class__)  # revealed: Literal[A]
+```
+
+## Inheritance (2)
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C(metaclass=A): ...
+class D(C, B): ...
+
+reveal_type(D.__class__)  # revealed: Literal[A]
+```
+
+## Inheritance (3)
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C(B): ...
+class D(metaclass=C): ...
+class E(D): ...
+
+reveal_type(E.__class__)  # revealed: Literal[C]
+```
+
+## Inheritance (4)
+
+```py
+class A: ...
+class B(metaclass=A): ...
+class C(metaclass=B): ...
+class D(C): ...
+
+reveal_type(D.__class__)  # revealed: Literal[B]
+```
```

Comment:
So we have two tests here for this kind of odd metaclass-with-a-metaclass scenario, whose value isn't entirely clear to me (the fact that a metaclass itself has a metaclass doesn't change anything about its compatibility with other metaclasses, and both of these tests are just simple linear inheritance, so no multiple base metaclasses to reconcile).

But it seems to me that we are missing tests for the scenarios that we are more likely to get wrong in the implementation where we have multiple different inherited metaclasses that _do_ have a subclass relationship and are thus compatible, and we should pick the child one. For instance a test for a case like this:

```py
class M(type): ...
class N(M): ...
class A(metaclass=M): ...
class B(metaclass=N): ...
class C(A, B): ...

reveal_type(C.__class__)  # revealed: Literal[N]
```

Or even a really complex case like this with diamond-inherited metaclasses, which actually should error:

```py
class M(type): ...
class M1(M): ...
class M2(M): ...
class M12(M1, M2): ...

class A(metaclass=M1): ...
class B(metaclass=M2): ...
class C(metaclass=M12): ...
class D(A, B, C): ...

# error: [conflicting-metaclass] "A and B have no subclass relationship"
reveal_type(D.__class__)
```

Reply:
Thanks, these are great!